### PR TITLE
Add mongoid support

### PR DESCRIPTION
**Pull request by ThibaudD**
### Generator

It is possible to generate a user class that uses mongoid with this command:

``` bash
rails g devise_token_auth:install User auth Mongoid
```
### Initializer settings

I added 2 initializer settings:
- add_mongoid_support, false by default
- use_only_mongoid, false by default, if you want to mix ActiveRecord object and Mongoid object

These are used in the method to determine if the concerned user class uses MongoId or ActiveRecord

``` ruby
def self.mongoid?(user_class)
    DeviseTokenAuth.add_mongoid_support &&
    ( DeviseTokenAuth.use_only_mongoid || user_class.included_modules.include?(Mongoid::Document) )
  end
```
### Tests

For the tests, I had to add FactoryGirl since Fixtures are into ActiveRecord.
I tried to reuse some tests like registrationsController and tokenValidationsController by wrapping them in a block, do you think it is ok doing this?

### Example
I made an example app with rails 5 with the api mode and mongoid, you can find it <a href="https://github.com/ThibaudD/tokenAuthMongoApi">here</a>

### CODE REVIEW

**Comment by MitchWeiss** [comment]
This is exactly what I'm looking for! What are the chances of this PR being accepted?

File: .travis.yml [hunk 1]
```diff
@@ -11,13 +11,16 @@ env:
   - DB=mysql
   - DB=postgresql
 
+services:
+  - mongodb
+
 script:
   - RAILS_ENV=test bundle exec rake --trace db:migrate
-  - RAILS_ENV=test CODECLIMATE_REPO_TOKEN=44d7688de8e1b567b4af25ec5083c2cc0a355ab911192a7cbefd1ea25b2ffd3d bundle exec rake
+  - RAILS_ENV=test bundle exec rake
```
Comment (booleanbetrayal) [inline .travis.yml]: did you mean to remove the `CODE_REPO_TOKEN` env variable assignment? I believe we need it for the build badge.

Comment (ThibaudD) [inline .travis.yml]: I removed it from my repository to not update your repository build badge when I push my commits on my repository. I re-added it.
We could add it in the Travic-ci env variables of this repository but it is an other issue.

**Comment by KRaymundus** [comment]
Great PR! Exactly what i was looking for.

Unfortunately, when testing with rspec I get this error: 
`ActiveRecord::ConnectionNotEstablished: No connection pool for ActiveRecord::Base`

In the PR there is an initializer in the gem, devise.rb that loads active_record if the `use_only_mongoid` boolean is false. Now here is the problem: this boolean is set in the initializer file, but according to what I read [here](http://www.sekuda.com/rails_initialization_and_configuration_order) the gems are initialized before the config/initializer/*.rb files. So the boolean is set after the gem initialized and the initializer in the gem has run.

Am I doing something wrong or is this an error that should be fixed before merging?

**Comment by ThibaudD** [comment]
Thank you for your comment, you are right.
Right now, I see two solutions, we can require the devise orm libs ourselves with a Devise initializer in our apps or in the engine.rb in the after_initialize block.
I will make the change with the second solution, be free to tell me if you think it is the right choice.

**Comment by filosofisto** [comment]
Does not works for me, look bellow:

rails g devise_token_auth:install User auth Mongoid
WARNING: Nokogiri was built against LibXML version 2.9.2, but has dynamically loaded 2.9.1
/usr/lib64/ruby/gems/2.1.0/gems/devise_token_auth-0.1.37/app/models/devise_token_auth/concerns/user.rb:103:in `tokens_has_json_column_type?': undefined method `table_exists?' for User:Class (NoMethodError)
        from /usr/lib64/ruby/gems/2.1.0/gems/devise_token_auth-0.1.37/app/models/devise_token_auth/concerns/user.rb:26:in `block in <module:User>'
        from /usr/lib64/ruby/gems/2.1.0/gems/activesupport-4.2.4/lib/active_support/concern.rb:120:in `class_eval'
        from /usr/lib64/ruby/gems/2.1.0/gems/activesupport-4.2.4/lib/active_support/concern.rb:120:in `append_features'
        from /home/eduardo/Private/CONSEG/conseg-web/app/models/user.rb:3:in `include'
        from /home/eduardo/Private/CONSEG/conseg-web/app/models/user.rb:3:in `<class:User>'
        from /home/eduardo/Private/CONSEG/conseg-web/app/models/user.rb:1:in `<top (required)>'
        from /usr/lib64/ruby/gems/2.1.0/gems/activesupport-4.2.4/lib/active_support/dependencies.rb:457:in `load'
        from /usr/lib64/ruby/gems/2.1.0/gems/activesupport-4.2.4/lib/active_support/dependencies.rb:457:in `block in load_file'
        from /usr/lib64/ruby/gems/2.1.0/gems/activesupport-4.2.4/lib/active_support/dependencies.rb:647:in `new_constants_in'
        from /usr/lib64/ruby/gems/2.1.0/gems/activesupport-4.2.4/lib/active_support/dependencies.rb:456:in `load_file'
        from /usr/lib64/ruby/gems/2.1.0/gems/activesupport-4.2.4/lib/active_support/dependencies.rb:354:in `require_or_load'
        from /usr/lib64/ruby/gems/2.1.0/gems/activesupport-4.2.4/lib/active_support/dependencies.rb:494:in `load_missing_constant'
        from /usr/lib64/ruby/gems/2.1.0/gems/activesupport-4.2.4/lib/active_support/dependencies.rb:184:in `const_missing'
        from /usr/lib64/ruby/gems/2.1.0/gems/activesupport-4.2.4/lib/active_support/inflector/methods.rb:261:in `const_get'
        from /usr/lib64/ruby/gems/2.1.0/gems/activesupport-4.2.4/lib/active_support/inflector/methods.rb:261:in `block in constantize'
        from /usr/lib64/ruby/gems/2.1.0/gems/activesupport-4.2.4/lib/active_support/inflector/methods.rb:259:in `each'
        from /usr/lib64/ruby/gems/2.1.0/gems/activesupport-4.2.4/lib/active_support/inflector/methods.rb:259:in `inject'
        from /usr/lib64/ruby/gems/2.1.0/gems/activesupport-4.2.4/lib/active_support/inflector/methods.rb:259:in `constantize'
        from /usr/lib64/ruby/gems/2.1.0/gems/activesupport-4.2.4/lib/active_support/dependencies.rb:566:in `get'
        from /usr/lib64/ruby/gems/2.1.0/gems/activesupport-4.2.4/lib/active_support/dependencies.rb:597:in `constantize'
        from /usr/lib64/ruby/gems/2.1.0/gems/devise-3.5.6/lib/devise.rb:292:in `get'
        from /usr/lib64/ruby/gems/2.1.0/gems/devise-3.5.6/lib/devise/mapping.rb:81:in `to'
        from /usr/lib64/ruby/gems/2.1.0/gems/devise-3.5.6/lib/devise/mapping.rb:76:in `modules'
        from /usr/lib64/ruby/gems/2.1.0/gems/devise-3.5.6/lib/devise/mapping.rb:93:in `routes'
        from /usr/lib64/ruby/gems/2.1.0/gems/devise-3.5.6/lib/devise/mapping.rb:160:in `default_used_route'
        from /usr/lib64/ruby/gems/2.1.0/gems/devise-3.5.6/lib/devise/mapping.rb:70:in `initialize'
        from /usr/lib64/ruby/gems/2.1.0/gems/devise-3.5.6/lib/devise.rb:326:in `new'
        from /usr/lib64/ruby/gems/2.1.0/gems/devise-3.5.6/lib/devise.rb:326:in `add_mapping'
        from /usr/lib64/ruby/gems/2.1.0/gems/devise-3.5.6/lib/devise/rails/routes.rb:238:in `block in devise_for'
        from /usr/lib64/ruby/gems/2.1.0/gems/devise-3.5.6/lib/devise/rails/routes.rb:237:in `each'
        from /usr/lib64/ruby/gems/2.1.0/gems/devise-3.5.6/lib/devise/rails/routes.rb:237:in `devise_for'
        from /home/eduardo/Private/CONSEG/conseg-web/config/routes.rb:9:in `block in <top (required)>'
        from /usr/lib64/ruby/gems/2.1.0/gems/actionpack-4.2.4/lib/action_dispatch/routing/route_set.rb:432:in `instance_exec'
        from /usr/lib64/ruby/gems/2.1.0/gems/actionpack-4.2.4/lib/action_dispatch/routing/route_set.rb:432:in `eval_block'
        from /usr/lib64/ruby/gems/2.1.0/gems/actionpack-4.2.4/lib/action_dispatch/routing/route_set.rb:410:in `draw'
        from /home/eduardo/Private/CONSEG/conseg-web/config/routes.rb:1:in `<top (required)>'
        from /usr/lib64/ruby/gems/2.1.0/gems/activesupport-4.2.4/lib/active_support/dependencies.rb:268:in `load'
        from /usr/lib64/ruby/gems/2.1.0/gems/activesupport-4.2.4/lib/active_support/dependencies.rb:268:in `block in load'
        from /usr/lib64/ruby/gems/2.1.0/gems/activesupport-4.2.4/lib/active_support/dependencies.rb:240:in `load_dependency'
        from /usr/lib64/ruby/gems/2.1.0/gems/activesupport-4.2.4/lib/active_support/dependencies.rb:268:in `load'
        from /usr/lib64/ruby/gems/2.1.0/gems/railties-4.2.4/lib/rails/application/routes_reloader.rb:40:in `block in load_paths'
        from /usr/lib64/ruby/gems/2.1.0/gems/railties-4.2.4/lib/rails/application/routes_reloader.rb:40:in `each'
        from /usr/lib64/ruby/gems/2.1.0/gems/railties-4.2.4/lib/rails/application/routes_reloader.rb:40:in `load_paths'
        from /usr/lib64/ruby/gems/2.1.0/gems/railties-4.2.4/lib/rails/application/routes_reloader.rb:16:in `reload!'
        from /usr/lib64/ruby/gems/2.1.0/gems/railties-4.2.4/lib/rails/application/routes_reloader.rb:26:in `block in updater'
        from /usr/lib64/ruby/gems/2.1.0/gems/activesupport-4.2.4/lib/active_support/file_update_checker.rb:75:in `call'
        from /usr/lib64/ruby/gems/2.1.0/gems/activesupport-4.2.4/lib/active_support/file_update_checker.rb:75:in `execute'
        from /usr/lib64/ruby/gems/2.1.0/gems/railties-4.2.4/lib/rails/application/routes_reloader.rb:27:in `updater'
        from /usr/lib64/ruby/gems/2.1.0/gems/railties-4.2.4/lib/rails/application/routes_reloader.rb:7:in `execute_if_updated'
        from /usr/lib64/ruby/gems/2.1.0/gems/railties-4.2.4/lib/rails/application/finisher.rb:69:in `block in <module:Finisher>'
        from /usr/lib64/ruby/gems/2.1.0/gems/railties-4.2.4/lib/rails/initializable.rb:30:in `instance_exec'
        from /usr/lib64/ruby/gems/2.1.0/gems/railties-4.2.4/lib/rails/initializable.rb:30:in `run'
        from /usr/lib64/ruby/gems/2.1.0/gems/railties-4.2.4/lib/rails/initializable.rb:55:in `block in run_initializers'
        from /usr/lib64/ruby/2.1.0/tsort.rb:226:in `block in tsort_each'
        from /usr/lib64/ruby/2.1.0/tsort.rb:348:in `block (2 levels) in each_strongly_connected_component'
        from /usr/lib64/ruby/2.1.0/tsort.rb:427:in `each_strongly_connected_component_from'
        from /usr/lib64/ruby/2.1.0/tsort.rb:347:in `block in each_strongly_connected_component'
        from /usr/lib64/ruby/2.1.0/tsort.rb:345:in `each'
        from /usr/lib64/ruby/2.1.0/tsort.rb:345:in `call'
        from /usr/lib64/ruby/2.1.0/tsort.rb:345:in `each_strongly_connected_component'
        from /usr/lib64/ruby/2.1.0/tsort.rb:224:in `tsort_each'
        from /usr/lib64/ruby/2.1.0/tsort.rb:205:in `tsort_each'
        from /usr/lib64/ruby/gems/2.1.0/gems/railties-4.2.4/lib/rails/initializable.rb:54:in `run_initializers'
        from /usr/lib64/ruby/gems/2.1.0/gems/railties-4.2.4/lib/rails/application.rb:352:in `initialize!'
        from /home/eduardo/Private/CONSEG/conseg-web/config/environment.rb:5:in `<top (required)>'
        from /usr/lib64/ruby/gems/2.1.0/gems/activesupport-4.2.4/lib/active_support/dependencies.rb:274:in `require'
        from /usr/lib64/ruby/gems/2.1.0/gems/activesupport-4.2.4/lib/active_support/dependencies.rb:274:in `block in require'
        from /usr/lib64/ruby/gems/2.1.0/gems/activesupport-4.2.4/lib/active_support/dependencies.rb:240:in `load_dependency'
        from /usr/lib64/ruby/gems/2.1.0/gems/activesupport-4.2.4/lib/active_support/dependencies.rb:274:in `require'
        from /usr/lib64/ruby/gems/2.1.0/gems/spring-1.6.4/lib/spring/application.rb:92:in `preload'
        from /usr/lib64/ruby/gems/2.1.0/gems/spring-1.6.4/lib/spring/application.rb:143:in `serve'
        from /usr/lib64/ruby/gems/2.1.0/gems/spring-1.6.4/lib/spring/application.rb:131:in `block in run'
        from /usr/lib64/ruby/gems/2.1.0/gems/spring-1.6.4/lib/spring/application.rb:125:in `loop'
        from /usr/lib64/ruby/gems/2.1.0/gems/spring-1.6.4/lib/spring/application.rb:125:in `run'
        from /usr/lib64/ruby/gems/2.1.0/gems/spring-1.6.4/lib/spring/application/boot.rb:18:in `<top (required)>'
        from /usr/lib64/ruby/site_ruby/2.1.0/rubygems/core_ext/kernel_require.rb:54:in `require'
        from /usr/lib64/ruby/site_ruby/2.1.0/rubygems/core_ext/kernel_require.rb:54:in `require'
        from -e:1:in `<main>'

**Comment by zachfeldman** [comment]
Hi there @ThibaudD    ,

In an effort to cleanup this project and prioritize a bit, we're marking pull requests that haven't had any activity in a while with a "close-in-7-days" label. If we don't hear from you in about a week, we'll be closing this pull request. Obviously feel free to re-open it at any time if it's the right time or this was done in error! If you do, please rebase it with the latest master and explain why it's still needed.

We really appreciate your contribution, we're just trying to make this project manageable again to move it forward.

Hope all is well.

**Comment by zachfeldman** [comment]
https://github.com/lynndylanhurley/devise_token_auth/issues/373 is related

**Pull request closed without merging**